nixos/modules/services/networking/nbd.nix

```
@@ -0,0 +1,152 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.nbd;
+  configFormat = pkgs.formats.ini { };
+  iniFields = with types; attrsOf (oneOf [ bool int float str ]);
+  serverConfig = configFormat.generate "nbd-server-config"
+    ({
+      generic = ({
+        user = "root";
+        group = "root";
+        port = cfg.server.listenPort;
+      }
+      // cfg.server.extraOptions
+      // (if (cfg.server.listenAddress != null)
+      then { listenaddr = cfg.server.listenAddress; }
+      else { }));
+    }
+    // (mapAttrs
+      (_: { path, allowAddresses, extraOptions }: {
+        exportname = path;
+      }
+      // extraOptions
+      // (if (allowAddresses != null)
+      then {
+        authfile = pkgs.writeText "authfile" ''
+          ${concatStringsSep "\n" allowAddresses}
+        '';
+      } else { }))
+      cfg.server.exports)
+    );
+  allowedDevices =
+    concatLists
+      (mapAttrsToList
+        (_: { path, ... }: if hasPrefix "/dev/" path then [ "${path} rw" ] else [ ])
+        cfg.server.exports);
+in
+{
+  options = {
+    services.nbd = {
+      server = {
```

not sure this extra level of attributes is helpful. how about this, matching the executable and service file names?
```suggestion
    services.nbd-server = {
```

---

I was thinking we'll probably also want `services.nbd.client`, but that would be a separate PR.  I think it would look cleaner with `services.nbd.*` instead of `services.nbd-server` and `server.nbd-client`.

---

that makes sense. wasn't aware you're planning on adding client services too.